### PR TITLE
Handle create buckets with common prefixes properly

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Handle create buckets with common prefixes properly

## Motivation and Context
common prefixes in bucket name if already created
are disallowed when etcd is configured due to the
prefix matching issue. Make sure that when we look
for bucket we are only interested in exact bucket
name, not the prefix.

## Regression
Maybe

## How Has This Been Tested?
Using etcd based configuration run etcd locally and then run minio server

```
~ MINIO_DOMAIN=localhost MINIO_ETCD_ENDPOINTS=http://localhost:2379 minio server ~/test
```

```
~ mc mb myminio/abcd1
...
```
Succeed

```
~ mc mb myminio/abcd
...
```
Should succeed ...

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by kannappanr — DISMISSED** [review]
LGTM. Tested the fix.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#1f93e891-19c1-488c-a86f-d8b068b23433) |
| mint-worm.sh | :x: [more...](#15b7f0db-63d2-45fb-9054-be609c691927) |
| mint-dist-xl.sh | :x: [more...](#431a59f9-39bf-4c6e-aba4-7973f8dcea38) |
<a name="1f93e891-19c1-488c-a86f-d8b068b23433"></a>
### 7556-f913abb/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:32234
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 15ae7f124d4a:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-jggpcv90zvdg6j",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 359,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.59:32234/aws-sdk-go-test-jggpcv90zvdg6j: dial tcp 72.28.97.59:32234: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="15b7f0db-63d2-45fb-9054-be609c691927"></a>
### 7556-f913abb/mint-worm.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:31095
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            worm
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ded3c0f0ce6b:/mint/log /tmp/mint-logs'

(1/1) Running worm tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "bucket-05f1jl9fqagrykcwo2uaaubq4le0e4gwlpiduqexfxq91f14op20ebg",
    "expiry": 60000000000,
    "objectName": "testObject"
  },
  "duration": 375,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.59:31095/bucket-05f1jl9fqagrykcwo2uaaubq4le0e4gwlpiduqexfxq91f14op20ebg: dial tcp 72.28.97.59:31095: connect: connection refused",
  "function": "PutAndDelete",
  "message": "WORM_MODE ON  - CreateBucket Failed",
  "name": "test worm mode",
  "status": "FAIL"
}

Executed 0 out of 1 tests successfully.
```
<a name="431a59f9-39bf-4c6e-aba4-7973f8dcea38"></a>
### 7556-f913abb/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:31172
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ef1ba49c0fc6:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 4 seconds
(5/14) Running awscli tests ... done in 1 minutes and 50 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 30 seconds
(8/14) Running minio-dotnet tests ... done in 35 seconds
(9/14) Running minio-go tests ... done in 53 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 32 seconds
(11/14) Running minio-js tests ... done in 38 seconds
(12/14) Running minio-py tests ... FAILED in 3 minutes and 9 seconds
{
  "args": {
    "bucket_name": "minio-py-test-a61302f8-53c3-4316-8c34-be99edb34e99",
    "object_name": "c21c232e-d2fc-49fd-b8e4-14de01427c2c"
  },
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1266, in test_presigned_get_object_response_headers\n    returned_content_language = response.headers['Content-Language']\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/_collections.py\", line 152, in __getitem__\n    val = self._container[key.lower()]\nKeyError: 'content-language'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1888, in main\n    test_presigned_get_object_response_headers(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1274, in test_presigned_get_object_response_headers\n    raise Exception(err)\nException: 'content-language'\n",
  "message": "'content-language'",
  "name": "minio-py:test_presigned_get_object_response_headers",
  "status": "FAIL",
  "duration": 108
}

Executed 11 out of 14 tests successfully.
```

**Review by kannappanr — APPROVED** [review]
LGTM

File: pkg/dns/etcd_dns.go [hunk 1]
```diff
@@ -70,7 +70,20 @@ func (c *coreDNS) Get(bucket string) ([]SrvRecord, error) {
 		if err != nil {
 			return nil, err
 		}
-		srvRecords = append(srvRecords, records...)
+		// Make sure we have record.Key is empty
+		// this can only happen when record.Key
+		// has bucket entry with exact prefix
+		// match any record.Key which do not
+		// match the prefixes we skip them.
+		for _, record := range records {
```
Comment (krisis) [inline pkg/dns/etcd_dns.go:78]: We can simplify the way we build srvRecords in `pkg/dns` package. For now, this change fixes the issue.

**Review by krisis — APPROVED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#342722dc-7f36-48e0-8c6c-7ac9219e447b) |
| mint-compression-dist-xl.sh | :x: [more...](#20c6505b-c425-4001-a991-c43e02c4cf6d) |
| mint-dist-xl.sh | :x: [more...](#d75ea4ea-59c7-4bd1-924e-b31aa09d2571) |
<a name="342722dc-7f36-48e0-8c6c-7ac9219e447b"></a>
### 7556-6159a5b/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:31260
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 45ad0047524f:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-2eslwklttz1zjq",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 287,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.59:31260/aws-sdk-go-test-2eslwklttz1zjq: dial tcp 72.28.97.59:31260: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="d75ea4ea-59c7-4bd1-924e-b31aa09d2571"></a>
### 7556-6159a5b/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32530
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp dbe455b02876:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-mg1s6j0jgg3mmw",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 380,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.55:32530/aws-sdk-go-test-mg1s6j0jgg3mmw: dial tcp 72.28.97.55:32530: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="20c6505b-c425-4001-a991-c43e02c4cf6d"></a>
### 7556-6159a5b/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:30104
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f4107c2eea60:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-w3pwgvn2tsys0c",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 335,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.53:30104/aws-sdk-go-test-w3pwgvn2tsys0c: dial tcp 72.28.97.53:30104: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```

**Pull request merged**